artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java

```diff
@@ -0,0 +1,648 @@
+ *
+ * @param <T> The type this class may hold, this is generic as can be anything that extends PriorityAware,
+ *         but intent is this is the QueueImpl:ConsumerHolder.
+ */
+public class QueueConsumersImpl<T extends PriorityAware> extends AbstractCollection<T> implements QueueConsumers<T> {
+
+   private final QueueConsumersIterator<T> iterator = new QueueConsumersIterator<>(this, true);
+
+   private volatile Level<T>[] levels;
+   private volatile int size;
+   private volatile T first;
+
+   private void setArray(Level<T>[] array) {
+      this.levels = array;
+   }
+
+   private Level<T>[] getArray() {
+      return levels;
+   }
+
+
+   public QueueConsumersImpl() {
+      levels = newLevelArrayInstance(0);
+   }
+
+   @SuppressWarnings("unchecked")
+   private static <T> Level<T>[] newLevelArrayInstance(int length) {
+      return (Level<T>[]) Array.newInstance(Level.class, length);
+   }
+
+   @Override
+   public int size() {
+      return size;
+   }
+
+   @Override
+   public boolean isEmpty() {
+      return size() == 0;
+   }
+
+   @Override
+   public Set<Integer> getPriorites() {
+      Level<T>[] levels = getArray();
+      return Arrays.stream(levels).map(Level::level).collect(Collectors.toSet());
+   }
+
+   @Override
+   public Iterator<T> iterator() {
+      return new QueueConsumersIterator<>(this, false);
+   }
+
+   @Override
+   public boolean hasNext() {
+      return iterator.hasNext();
+   }
+
+   @Override
+   public T next() {
+      return iterator.next();
+   }
+
+   @Override
+   public QueueConsumers<T> reset() {
+      iterator.reset();
+      return this;
+   }
+
+   @Override
+   public void forEach(Consumer<? super T> action) {
+      Objects.requireNonNull(action);
+      Level<T>[] current = getArray();
+      int len = current.length;
+      for (int i = 0; i < len; ++i) {
+         current[i].forEach(action);
+      }
+   }
+
+   private Level<T> getLevel(int level, boolean createIfMissing) {
+      Level<T>[] current = getArray();
+      int low = 0;
+      int high = current.length - 1;
+
+      while (low <= high) {
+         int mid = (low + high) >>> 1;
+         Level<T> midVal = current[mid];
+
+         if (midVal.level() > level)
+            low = mid + 1;
+         else if (midVal.level() < level)
+            high = mid - 1;
+         else
+            return midVal; //key found
+      }
+
+      if (createIfMissing) {
+         Level<T>[] newLevels = newLevelArrayInstance(current.length + 1);
+         if (low > 0) {
+            System.arraycopy(current, 0, newLevels, 0, low);
+         }
+         if (current.length - low > 0) {
+            System.arraycopy(current, low, newLevels, low + 1, current.length - low);
+         }
+         newLevels[low] = new Level<T>(level);
+         setArray(newLevels);
+         return newLevels[low];
+      }
+      return null;
+   }
+
+   @Override
+   public synchronized boolean add(T t) {
```

Comment:
it could happen that many will access this one concurrently?

Reply:
Yes, its following the lines on design of CopyOnWriteArrayList, but using syncronized methods over using Reentrant lock as concurrency expected is low.

Reply:
I understand it, but my point is more an idea to improve/simplify things: I understand that `CopyOnWriteArrayList` is the original starting point but here we are doing something similar, but different.
If we just use few methods of it I would instead just implement what we need, avoiding to test/provide behaviours for more complex things that we are not using from it. The same to be said from the concurrency behaviour: if we have this class designed to be used in a single writer fashing, making the code simpler (and faster most of the time), I would do this instead.
We are not implementing here any new CopyOnWriteArrayList, but a specific domain -specific class so I suppose that won't hurt to make it simpler (if possible)

Reply:
All methods added we are using. Its not used in single write. This class essentially replaces the CopyOnWriteArrayList we used to hold the consumers, and supported only simple round robin before, and why most of its design is based on that.

Reply:
I assume thumbs up, means your query is resolved, ill mark this as resolved so i can keep track of open things i need to address.